punishments.js

```diff
@@ -667,18 +683,30 @@ Punishments.roomBan = function (room, user, expireTime, id, ...rest) {
 };
 
 Punishments.roomBlacklist = function (room, user, expireTime, id, ...rest) {
-	if (!id) id = user.getLastId();
+	if (!id && user) id = user.getLastId();
 
 	if (!expireTime) expireTime = Date.now() + BLACKLIST_DURATION;
 	let punishment = ['BLACKLIST', id, expireTime].concat(rest);
-	Punishments.roomPunish(room, user, punishment);
+	if (user) {
```

Comment:
Would definitely attempt to extract `user` from `id` before going down this road.